tuplex/codegen/src/FunctionRegistry.cc

```diff
@@ -966,6 +966,301 @@ namespace tuplex {
+
+                return SerializableValue(_env.boolConst(false), _env.i64Const(sizeof(int64_t)));
+            }
+        }
+
+        codegen::SerializableValue FunctionRegistry::createMathIsCloseCall(llvm::IRBuilder<>& builder, const python::Type &argsType,
+                                                     const python::Type &retType,
+                                                     const std::vector<tuplex::codegen::SerializableValue> &args) {
+            assert(argsType.isTupleType());
+            assert(args.size() == argsType.parameters().size());
+            assert(args.size() >= 2);
+
+            using namespace llvm;
+            auto& context = builder.GetInsertBlock()->getContext();
+            Module *M = builder.GetInsertBlock()->getModule();
+            std::vector<python::Type> input_types = argsType.parameters();
+
+            auto x_val = args[0];
+            auto y_val = args[1];
+            llvm::Value* rel_tol_val;
+            llvm::Value* abs_tol_val;
+            auto x_ty = input_types[0];
+            auto y_ty = input_types[1];
+
+            if (args.size() == 2) {
+                // rel_tol and abs_tol not specified
+                rel_tol_val = _env.f64Const(1e-09);
+                abs_tol_val = _env.f64Const(0);
+            } else if (args.size() == 3) {
+                // assume that the third argument is rel_tol
+                rel_tol_val = args[2].val;
+                abs_tol_val = _env.f64Const(0);
+            } else {
+                assert(args.size() == 4);
+                // assume that the third argument is rel_tol and the fourth argument is abs_tol
+                // this means we don't handle the case where abs_tol is specified but rel_tol is not
+                rel_tol_val = args[2].val;
+                abs_tol_val = args[3].val;
+            }
+
+            auto rel_tol = _env.upCast(builder, rel_tol_val, _env.doubleType());
+            auto abs_tol = _env.upCast(builder, abs_tol_val, _env.doubleType());
+
+            // check x and y types - bools and ints can be optimized!
+            // TODO: error check both rel_tol > 0 and abs_tol >= 0
+            if (x_ty == python::Type::BOOLEAN && y_ty == python::Type::BOOLEAN) {
+                auto xor_xy = builder.CreateXor(x_val.val, y_val.val);
+                auto rel_cmp = builder.CreateFCmpOGE(rel_tol, _env.f64Const(1));
+                auto abs_cmp = builder.CreateFCmpOGE(abs_tol, _env.f64Const(1));
+                auto rel_or_abs = builder.CreateOr(rel_cmp, abs_cmp);
+                auto eq_check = builder.CreateXor(xor_xy, _env.boolConst(true));
+                auto or_res = builder.CreateOr(rel_or_abs, eq_check);
+
+                auto resVal = _env.upcastToBoolean(builder, or_res);
+                auto resSize = _env.i64Const(sizeof(int64_t));
+
+                return SerializableValue(resVal, resSize);
+            } else if (x_ty != python::Type::F64 && y_ty != python::Type::F64) {
```

Comment:
what about the case x = F64, y=I64? this would be done here as well? why do we then carry out a cast to integers?